pyccel/ast/basic.py

```diff
@@ -183,8 +199,10 @@ def substitute(self, original, replacement, excluded_nodes = ()):
 
         def prepare_sub(found_node):
             idx = original.index(found_node)
-            found_node.remove_user_node(self)
-            replacement[idx].set_current_user_node(self)
+            if not self.ignore(found_node):
+                found_node.remove_user_node(self)
+            if not self.ignore(replacement[idx]):
+                replacement[idx].set_current_user_node(self)
             return replacement[idx]
```

Comment:
I wonder if the definition of this function inside the `substitute` method could slow down the Python interpreter, especially when `substitute` is called recursively. This deserves a bit of profiling on a complex example.

If it does, you could define the method (where I reduced the number of calls to `__getitem__`from 3 to just 1)
```python
def prepare_substitution(self, original, replacement, node):
    idx = original.index(node)
    rep = replacement[idx]
    if not self.ignore(node):
        node.remove_user_node(self)
    if not self.ignore(rep):
        rep.set_current_user_node(self)
    return rep
```
and then use a lambda function for clarity:
```python
prepare_sub = lambda found_node: self.prepare_substitution(original, replacement, found_node)
```

Reply:
I tried to do some simple profiling using the `test_substitution` method. As a rule of thumb it seems that defining a `prepare_substitution(self, original, replacement, node)` with a lambda function causes a slowdown (3rd significant digit), I suspect that creating a lambda function is not much cheaper than creating a function.
Using the function directly instead of via a lambda function seems to cause a slight speedup (4th significant digit).

But these results should be taken with a grain of salt as my computer is not very reliable. The results were found by running the test 1000 times using timeit. Each set of 1000 took around 10s (but closing internet made me gain a second, which shows how trustworthy this is :/ )

Reply:
For now I have just fixed the `__getitem__` calls (also 4th significant difference). See 4e00d21